var/spack/repos/builtin/packages/highfive/package.py

```diff
@@ -45,8 +45,9 @@ class Highfive(CMakePackage):
     variant('mpi', default=True, description='Support MPI')
 
     depends_on('boost @1.41:', when='+boost')
-    depends_on('hdf5 ~mpi', when='~mpi')
-    depends_on('hdf5 +mpi', when='+mpi')
+    depends_on('hdf5',         when='~mpi')
+    depends_on('hdf5+mpi',     when='+mpi')
+    depends_on('mpi',          when='^hdf5+mpi')
```

Comment:
I would make this `depends_on('mpi',          when='+mpi')`

Reply:
It requires mpi if the hdf5 version was compiled with mpi, even if we didn't ask for +mpi.
Again the idea is that it can link to whatever version (hdf5+mpi/hdf5~mpi) it found
We can discuss this better offline, because I think it makes sense but I am still not convinced Spack supports it 100%